stdlib/array.ncl

```diff
@@ -346,5 +408,200 @@
+
+          # Preconditions
+
+          `n` must be an integer greater or equal to `0`.
+
+          # Examples
+
+          ```nickel
+          replicate 0 false
+            => [ ]
+          replicate 5 "x"
+            => [ "x", "x", "x", "x", "x" ]
+          ```
+        "%
+      = fun n x => %generate% n (fun _i => x),
+
+    range_step
+      : Number -> Number -> Number -> Array Number
+      | doc m%"
+          `range_step start end step` generates the array of numbers
+          `[start, start + step, start + 2*step, ..]` up to the first element
+          (excluded) larger than or equal to `end`
+
+          # Preconditions
+
+          In `range_step start end step`, `start` and `end` must verifies `start
+          <= end`.`step` must be strictly greater than `0` (or the function will
+          loop).
+
+          # Examples
+
+          ```nickel
+          range_step (-1.5) 2 0.5
+           => [ -1.5, -1, -0.5, 0, 0.5, 1, 1.5 ]
+          ```
+        "%
+      = fun start end step => %generate%
+          ((end - start) / step |> number.floor)
+          (fun i => start + i*step),
+
+    range
+      : Number -> Number -> Array Number
+      | doc m%"
+          `range start end` generates the array of numbers
+          `[start, start + 1, start + 2, ..]` up to the first element
+          (excluded) larger than or equal to `end`.
+
+          `range start end` is equivalent to `range_step start end 1`.
+
+          # Preconditions
+
+          In `range_step start end`, `start` and `end` must verify
+          `start <= end`.
+
+          # Examples
+
+          ```nickel
+          range 0 5
+           => [ 0, 1, 2, 3, 4 ]
+          ```
+        "%
+      = fun start end => range_step start end 1,
+
+    reduce_left
+      : forall a. (a -> a -> a) -> Array a -> a
+      | doc m%"
+        Reduces the elements to a single one, by repeatedly applying a reducing
+        operation. If the array is empty, returns an empty array.
+
+        `reduce_left` associates to the left, that is
+        `reduce_left op [x1, x2, ..., xn]` results in `op (... (op (op x1 x2) x3) ...) xn`.
+
+        `reduce_left` is the same as `fold_left`, but uses the first element as
+        the initial element.
+
+        # Left vs right
+
+        The rationale to decide between `fold_left` and `fold_right` applies to
+        `reduce_left` and `reduce_right` as well. See the documentation of
+        `fold_left`.
+
+        # Examples
+
+        ```nickel
+          reduce_left (@) [ [1, 2], [3], [4,5] ]
+            => (([1, 2] @ [3]) @ [4,5])
+            => [ 1, 2, 4, 5 ]
+          reduce_left (-) [ 1, 2, 3, 4]
+            => ((1 - 2) - 3) - 4
+            => -8
+        ```
+        "%
+      = fun f array =>
+        if array == [] then
+          array
+        else
+          let first = %elem_at% array 0 in
+          let rest = %array_slice% 1 (%length% array) array in
+          fold_left f first rest,
```

Comment:
I don't think returning an empty array when given one makes sense here (How does this even typecheck?). It would probably be best to somehow fail on an empty array; maybe by applying the `array.NonEmpty` contract somewhere?

Reply:
It doesn't make any sense, I think I thought about returning `None` and may brain then failed. We should fail on this, probably.

The issue I had with more specific contracts like `NonEmpty` is that they make those standard functions a pain to use in statically typed code. On one hand it's a shame, because we have those nice contracts with argument tracking and whatnot which could provide really good error messages. On the other hand, making folds, reduce and map a pain to use in typed code is also kind of a show stopper. One solution could be to separate them in submodules (having a more loosely typed version, and a tight contract-protected version), but that isn't satisfying either.

Maybe one solution at some point would be the ideas exposed in #420.

Reply:
Would that work? Now `array` has type `NonEmpty`, and you can't do anything with it in typed code anymore, I fear.

Reply:
Or maybe something like `%seq% (array | NonEmpty) ...restofthefunction...` but this starts to be really convoluted :sweat_smile:

Reply:
Maybe this calls for a construct to check a contract without influencing typechecking

Reply:
So, we'd write
```
reduce_left
  : forall a. (a -> a -> a) -> Array a -> a
  | forall a. (a -> a -> a) -> NonEmpty -> a
  = ...
```
I guess? That's a little bit of duplication, but I think it's a price worth paying here 👍

Reply:
We can also just add the "differential" contract, so to speak:

```
reduce_left
  : forall a. (a -> a -> a) -> Array a -> a
  | Dyn -> NonEmpty -> Dyn
```

That might make for a slightly worse error message, though. We could also duplicate the type fully as you propose, and rely on contract elision (yet to be added) to not do too much work.